2000-01-27  Jakub Jelinek  <jakub@redhat.com>

	* reload1.c (emit_input_reload_insns): Honour forcing of constants
	into memory by PREFERRED_RELOAD_CLASS NO_REGS.

--- gcc/reload1.c.jj	Tue Jan 18 22:33:00 2000
+++ gcc/reload1.c	Thu Jan 27 19:26:44 2000
@@ -6326,7 +6326,10 @@ emit_input_reload_insns (chain, rl, old,
 	      && ((reg_equiv_memory_loc
 		   [REGNO (SUBREG_REG (oldequiv))] != 0)
 		  || (reg_equiv_constant
-		      [REGNO (SUBREG_REG (oldequiv))] != 0))))
+		      [REGNO (SUBREG_REG (oldequiv))] != 0)))
+	  || (CONSTANT_P (oldequiv)
+	      && PREFERRED_RELOAD_CLASS (oldequiv,
+			REGNO_REG_CLASS (REGNO (reloadreg))) == NO_REGS))
 	real_oldequiv = rl->in;
       gen_reload (reloadreg, real_oldequiv, rl->opnum,
 		  rl->when_needed);
